thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+of doing the things I was stumbling with.
+
+This is more of a sketch of my experiences with the language, specifically from the point of
+view of someone coming from Rust; used to the Rusty way of doing things. It might be useful as an
+advisory to Rustaceans thinking about trying the language out, and what to expect.
+
+[raft]: https://github.com/Manishearth/cs733/tree/master/assignment3/raft
+[dash]: http://manishearth.github.io/dash
+
+// I've structured this as "what I liked"/"what I didn't like", but I'm open to other ways of structuring this too
+
+## What I liked
+
+Despite the performance costs, having a GC at your disposal after using Rust for very long is quite liberating. 
+For a while my internalized borrow checker would throw red flags on me tossing around data indiscriminately,
+but I learned to  ignore it as far as Go code goes. I was able to quickly share state via pointers without
+worrying about safety, which was quite useful.
+
+Having channels as part of the language itself was also quite ergonomic.  The `data <- chan` and `chan <- data` syntax is fun to use,
+and whilst it's not very different from `.send()` and `.recv()` in Rust, I found it surprisingly easy to read.
+Initially I got confused often by which side the channel was, but after a while I got used to it.
+It also has an in built select block for selecting over channels (Rust has a macro for the same purpose, but no language support).
+
+
+`gofmt`. The Go style of coding is different from the Rust one (tabs vs spaces, how declarations look),
+but I continued to use the Rust style because of the muscle memory (also too lazy to change the settings
+in my editor). `gofmt` made life easy since I could just run it in a directory and it would fix everything.
+Eventually I was able to learn the proper style by watching my code get corrected.
+I'd love to see a `rustfmt`, however!
+
+
+Go is great for debugging programs with multiple threads, too. It can detect deadlocks and
+post traces for the threads (with metadata including what code the thread was spawned from, as well as its
+current state). It also posts such traces when the program crashes. These are totally awesome and saved me tons of time whilst debugging
+my code (which at times had all sorts of cross interactions between more than ten goroutines in the tests)
+Without a green threading framework, I'm not sure how easy it will be to integrate this into Rust
+(for debug builds, obviously), but I'd certainly like it to be there some day.
+
+Go has really great <s>green threads</s> goroutines. They're rather efficient
+(I can spawn a thousand and it schedules them nicely), and easy to use.
+
+Go has really good in built support and tooling for tests (Rust does too).
+I enjoyed writing tests in Go quite a bit due to this.
+
+## What I didn't like
+
+Sadly, there are a lot of things here, but bear in mind what I mentioned above
```

Comment:
I might rephrase this paragraph slightly, to make it less likely that you get negative feedback from Go enthusiasts, perhaps:

> There are quite a few things here, but bear in mind I'm new to Go, and am still learning the _Go way_ of doing things. I would love to hear feedback about how I can overcome some of the problems I ran into.